### PR TITLE
[BugFix][Scheduler] Return per-request spec decode metrics

### PR DESCRIPTION
**Pull request by Big2Wheel**
### What this PR does / why we need it?

Upstream vLLM added per-request speculative decoding metrics in vllm-project/vllm#48915. The Ascend `RecomputeScheduler` maintains its own `update_from_output` implementation, but it did not update `RequestSpecDecodeMetrics` or attach the collected metrics to the final `EngineCoreOutput`. Consequently, requests using the recompute scheduler could not return the requested speculative decoding metrics.

This PR keeps the Ascend scheduler aligned with the upstream behavior by:

- recording drafted and accepted tokens for each request;
- excluding grammar-invalid draft tokens from the drafted-token count;
- attaching metrics only to the request's final engine output;
- adding focused unit coverage for the recompute scheduler;
- adding a single-card NPU OpenAI API E2E test for normal and streaming responses;
- documenting the `summary` and `detailed` response formats.

Related upstream change: https://github.com/vllm-project/vllm/pull/48915

### Does this PR introduce _any_ user-facing change?

Yes. When `--per-request-spec-decode-metrics summary` or `detailed` is explicitly enabled, requests handled by the Ascend recompute scheduler now return `metrics.speculative_decoding` in the final response. The default `none` behavior is unchanged.

### How was this patch tested?

- Unit regression: `pytest -q tests/ut/core/test_recompute_scheduler.py` (`6 passed`).
- Single-card Ascend 910B3 E2E with Qwen3-0.6B and the n-gram proposer: `VLLM_TEST_MODEL=Qwen/Qwen3-0.6B pytest -sv tests/e2e/pull_request/one_card/spec_decode/test_per_request_metrics.py` (`1 passed`).
- Manually verified `none`, `summary`, and `detailed` modes; streaming only exposes metrics on the final usage chunk; `n=2` remains unsupported and does not expose misleading metrics; n-gram and suffix proposer stability loops passed.
- Ruff 0.14.0 formatting and lint checks, Markdownlint 0.45.0, Codespell 2.4.1, and `git diff --check` passed for the changed files.
- The full `format.sh ci` wrapper was also attempted, but its first-time Actionlint bootstrap could not download a Go dependency because `proxy.golang.org` timed out. The direct language-specific checks above passed.

#### Real MTP model validation on Ascend 910B3

The OpenAI-compatible response path was additionally validated with two real Qwen3.6 checkpoints on two Ascend 910B3 64 GiB NPUs. Both services were restricted to physical devices 4 and 5 with `ASCEND_RT_VISIBLE_DEVICES=4,5` and used three speculative tokens per MTP step.

Common relevant arguments:

```bash
--tensor-parallel-size 2 \
--speculative-config '{"method":"qwen3_5_mtp","num_speculative_tokens":3,"enforce_eager":true}' \
--compilation-config '{"cudagraph_mode":"FULL_DECODE_ONLY"}' \
--per-request-spec-decode-metrics detailed
```

Model-specific configurations and results:

| Model | Configuration | Detailed response | Streaming | Stability |
| --- | --- | --- | --- | --- |
| Qwen3.6-27B-w8a8 | TP2, Ascend W8A8 | Passed | Passed | 5 sequential + 4 concurrent requests passed |
| Qwen3.6-35B-A3B | TP2 + EP2, BF16 MoE | Passed | Passed | 5 sequential + 4 concurrent requests passed |

Representative 27B response metrics:

```json
{
  "mean_acceptance_length": 3.6666666666666665,
  "draft_acceptance_rate": 0.8888888888888888,
  "num_spec_steps": 18,
  "num_accepted_draft_tokens": 48,
  "num_draft_tokens": 54,
  "num_spec_tokens": 3
}
```

Representative 35B-A3B response metrics:

```json
{
  "mean_acceptance_length": 3.142857142857143,
  "draft_acceptance_rate": 0.7142857142857143,
  "num_spec_steps": 21,
  "num_accepted_draft_tokens": 45,
  "num_draft_tokens": 63,
  "num_spec_tokens": 3
}
```

For both models, the following invariants were checked:

- the acceptance histogram sums to `num_spec_steps`;
- its weighted sum equals `num_accepted_draft_tokens`;
- the per-step accepted and drafted arrays reproduce their aggregate totals;
- every step satisfies `0 <= accepted <= drafted`;
- the reported acceptance rate and mean acceptance length match the aggregate counts;
- in streaming mode, metrics appear exactly once, on the final usage chunk.

No `ERROR`, traceback, OOM, or runtime failure was observed in either model log. This online-serving validation exercises the standard scheduler's upstream metric propagation with the Ascend MTP model runner and backend. The Ascend-specific `RecomputeScheduler` is selected only on a PD-disaggregated decode node, so its changed path is covered by the focused unit regression above rather than by these standalone `vllm serve` runs.

- vLLM main: https://github.com/vllm-project/vllm/commit/ba07e4a48fc951300d97eb506217dd530583dea3

### CODE REVIEW

File: vllm_ascend/core/recompute_scheduler.py [hunk 1]
```diff
@@ -1189,6 +1189,18 @@ def update_from_output(
                     num_invalid_spec_tokens=scheduler_output.num_invalid_spec_tokens,
                     request_id=req_id,
                 )
+                if request.spec_decode_metrics is not None:
+                    # Exclude grammar-invalidated drafts from the proposed
+                    # count, mirroring make_spec_decoding_stats; the accepted
+                    # bucket (j) is unaffected.
+                    adj_draft_tokens = num_draft_tokens
+                    if scheduler_output.num_invalid_spec_tokens:
+                        adj_draft_tokens -= scheduler_output.num_invalid_spec_tokens.get(req_id, 0)
```
[the remark this answers is not in the document]
Comment (Big2Wheel) [reply vllm_ascend/core/recompute_scheduler.py:1198]: Thanks for checking this invariant. I compared this path with the implementation introduced by upstream vLLM PR #48915: the upstream scheduler intentionally passes the grammar-adjusted draft count directly, without clamping it to the accepted count. The structured-output contract guarantees that grammar-invalidated proposals are excluded before verification, so num_accepted cannot exceed num_draft_tokens minus the invalid count. Clamping here would hide a scheduler/accounting bug and make Ascend metrics diverge from upstream and from the aggregate make_spec_decoding_stats counters. For those reasons I would prefer to keep the exact upstream behavior. The regression test also covers the boundary case of 3 drafts, 1 invalid draft, and 2 accepted drafts.

**Comment by Big2Wheel** [comment]
The initial CI checks are clean: pre-commit, CPU UT, documentation builds, link check, and DCO all passed. The ci-gate is currently blocked only because this source-code PR requires a maintainer-applied test label. Could a maintainer please add the recommended ready-precise label? The added one-card E2E was also run locally on an Ascend 910B3 with Qwen3-0.6B and passed.

**Comment by Big2Wheel** [comment]
Added real-model MTP validation details to the PR description.

Validated on two Ascend 910B3 64 GiB NPUs (`ASCEND_RT_VISIBLE_DEVICES=4,5`):

- Qwen3.6-27B-w8a8: TP2, Ascend W8A8;
- Qwen3.6-35B-A3B: TP2 + EP2, BF16 MoE;
- MTP with 3 speculative tokens and `FULL_DECODE_ONLY`;
- `--per-request-spec-decode-metrics detailed` for non-streaming and streaming OpenAI requests;
- all aggregate/per-step metric invariants passed;
- metrics appeared exactly once in the final streaming usage chunk;
- 5 sequential and 4 concurrent requests passed for each model;
- no error, traceback, OOM, or runtime failure was found in either log.

For clarity, standalone `vllm serve` uses the standard scheduler, so these runs validate the real Ascend MTP/API integration. The PR's Ascend-specific `RecomputeScheduler` path is selected only for a PD-disaggregated decode node and is covered by the focused scheduler unit regression.
